<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Release History &mdash; gemini 0.18.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/labibi.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.18.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/gemini.png"/>
    <link rel="top" title="gemini 0.18.1 documentation" href="../index.html" />
    <link rel="next" title="F.A.Q." href="faq.html" />
    <link rel="prev" title="&lt;no title&gt;" href="caveats.html" />


<script src="../_static/labibi.js"></script>



<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', '']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>


  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="faq.html" title="F.A.Q."
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="caveats.html" title="&lt;no title&gt;"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">gemini v0.18.1</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  
  <div class="section" id="release-history">
<h1>Release History<a class="headerlink" href="#release-history" title="Permalink to this headline">¶</a></h1>
<div class="section" id="future">
<h2>0.19.0 (future)<a class="headerlink" href="#future" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Use vcfanno for faster, more generalized variant annotation and database creation.</li>
<li>Use SQLAlchemy for table definitions in an effort to support different RDBMS backends.</li>
<li>X-linked recessive and dominant tools.</li>
</ol>
</div>
<div class="section" id="id1">
<h2>0.18.1<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Add <code class="docutils literal"><span class="pre">clinvar_gene_phenotype</span></code> column which can be used to limit candidates to those that are in the same
gene as a gene with the known phenotype from clinvar. Phenotypes are all lower case.
Likely usage is: <code class="docutils literal"><span class="pre">--filter</span> <span class="pre">&quot;...</span> <span class="pre">and</span> <span class="pre">clinvar_gene_phenotype</span> <span class="pre">LIKE</span> <span class="pre">'%dysplasia%'</span></code> where the &#8216;%&#8217; wildcard is needed
because the column is a &#8216;|&#8217; delimited list of all disease for that gene.</li>
<li>Add <code class="docutils literal"><span class="pre">geno2mp_hpo_ct</span></code> column which will be &gt; 0 if that variant is present in geno2mp (<a class="reference external" href="http://geno2mp.gs.washington.edu/Geno2MP/#/">http://geno2mp.gs.washington.edu/Geno2MP/#/</a>)</li>
<li>Fix <code class="docutils literal"><span class="pre">comp_hets</span></code> tool when <code class="docutils literal"><span class="pre">exac_num_hom_alt</span></code> was requested. (Thanks to &#64;davemcg for reporting).</li>
<li>Change <code class="docutils literal"><span class="pre">splice_region_variant</span></code> from <code class="docutils literal"><span class="pre">LOW</span></code> to <code class="docutils literal"><span class="pre">MED</span></code> priority in the interest of reducing false negatives in the study of rare disease.</li>
<li>Set missing AF to -1 (instead of NULL) for unknown for all ESP, 1KG, ExAC allele frequency columns.</li>
<li>Don&#8217;t user <code class="docutils literal"><span class="pre">order</span> <span class="pre">by</span></code> when not needed for built in tools. Speeds up queries when min-kindreds is None or 1.</li>
<li>Document the <code class="docutils literal"><span class="pre">--min-gq</span></code> option</li>
<li>Fix gemini load with <code class="docutils literal"><span class="pre">-t</span> <span class="pre">all</span></code> when only VEP is present.</li>
<li>Fix bug in gemini <code class="docutils literal"><span class="pre">annotate</span></code> where numeric operations on integers did not work for VCF.</li>
<li>Fix bug in comp_hets tool where attempting to phase a &#8221;.&#8221; genotype would result in an error. (Thanks to Aparna and Martina for reporting)</li>
<li>Fix bug in <code class="docutils literal"><span class="pre">mendel_errors</span> <span class="pre">tool</span></code> when specifying <code class="docutils literal"><span class="pre">--families</span></code>. Thanks to &#64;davemcg for reporting.</li>
</ol>
</div>
<div class="section" id="id2">
<h2>0.18.0<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple" start="0">
<li>Improved installation and update via conda (via Brad Chapman!).</li>
<li>Update support for SO term variant impact predictions via VEP and SnpEff and support newest snpEff version.
If both snpEff and VEP annotations are present use <cite>gemini load -t all</cite> to save all annotations and the most
deleterious will be saved in the variants table.</li>
<li>Add an <cite>is_splicing</cite> column.</li>
<li>Add <cite>exac_num_het</cite>, <cite>exac_num_hom_alt</cite> and <cite>exac_num_chroms</cite> columns. (See #568).</li>
<li>Fix issues with cyvcf2 handling of haploid calls (from X chromosome from GATK; thanks Athina for reporting)</li>
<li>Fix handling of VEP extra fields (thanks &#64;jsh58).</li>
<li>Remove pygraph dependency (networkx is easier to install). Allow specifying custom edges to interactions tools.</li>
<li>Fix some edge-cases in compound het tool. Thanks to Jamie Kwok for reporting.</li>
<li>Allow requring a minimum genotype quality in inheritance tools with &#8211;min-gq option.</li>
<li>Fix bug in cyvcf2 for parsing hemizyous (0/.) variants from decomposition.</li>
<li><cite>comp_het</cite> tool now has a priority 2 level for unphased singletons where both sites are HET. Candidates where both parents and HET, along with the affected are now priority 3.</li>
</ol>
</div>
<div class="section" id="id3">
<h2>0.17.2<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Fix bcolz dependencies owing to issues with bcolz 0.11.0</li>
</ol>
</div>
<div class="section" id="id4">
<h2>0.17.1<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Change handling of missing values in PL/GL (&#8211;gt-pl-max) so that missing values are set to int32 max. (Thanks Karyn for reporting).</li>
<li>Fix distributed loading of VEP with extra columns (&#64;chapmanb) [Regression since 0.17.0]</li>
<li>Fix comp_het test and improve efficiency (thanks Bianca for reporting)</li>
<li>Bug fix: populate eval dictionary with sample_info.</li>
</ol>
</div>
<div class="section" id="id5">
<h2>0.17.0<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>switch to cyvcf2 to speed loading</li>
<li>per-sample depths are calculated using AD (GATK) or AO+RO (Freebayes). This makes depth filters more conservative.</li>
<li>extra VEP annotations are loaded with loading machinery, not as an extra step as before.</li>
<li>add max_aaf_all column (<a class="reference external" href="https://github.com/arq5x/gemini/issues/520">https://github.com/arq5x/gemini/issues/520</a>) as an aggregate of a number of population filters.</li>
<li>use &#8211;families to limit queries <em>before</em> any work is done. Thanks to Bianca for reporting.</li>
<li>No longer create bcolz indicies by default. Users can create them with <cite>gemini bcolz_index</cite>.</li>
<li>New <cite>genewise</cite> tool. See docs.</li>
<li>gemini load: &#8211;skip-info-string has been replaced with &#8211;save-info-string and the INFO field is not longer saved by default.</li>
<li>comp_hets: default to only showing confident (priority 1) candidates. Show all candidates with &#8211;max-priority 3.</li>
</ol>
</div>
<div class="section" id="id6">
<h2>0.16.3<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Fix bug in <code class="docutils literal"><span class="pre">comp_het</span></code> with reporting same pair multiple times.</li>
<li>Handle UNKOWN genotypes in <code class="docutils literal"><span class="pre">comp_het</span></code> tool</li>
<li>Fix cyvcf dependency in requirements</li>
<li>Only run tests that require bgzip/tabix/bedtools if they are available on PATH</li>
<li>Limit ipython version to 3&lt;version&lt;4</li>
</ol>
</div>
<div class="section" id="id7">
<h2>0.16.2<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Hone rules for unphased and partially-phased compound hets.</li>
<li>Remove <cite>&#8211;lenient</cite> argument for comp_hets and add <cite>&#8211;pattern-only</cite> to find compound_hets regardless of affection status.</li>
<li>The <cite>&#8211;lenient</cite> argument to the <cite>autosomal_dominant</cite> tool has been relaxed to allow parents with unknown phenotypes.</li>
<li>Re (vt) decompose data files for 1000 genomes and ExAC (thanks Julien and Xiaolin for reporting).</li>
</ol>
</div>
<div class="section" id="id8">
<h2>0.16.1<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Fix regression in loading when AAF is None</li>
<li>Fix handing in mendelian error tool where all genotype likelihoods are low (thanks Bianca)</li>
<li>Don&#8217;t phase de-novo&#8217;s (caused error in comp_het tool). (thanks Bianca)</li>
<li>Fix regression in loading VEP with multicore (thanks Andrew)</li>
</ol>
</div>
<div class="section" id="id9">
<h2>0.16.0<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>The built-in inheritance model tools (<code class="docutils literal"><span class="pre">auto_rec</span></code>, etc.) have been modified to be more
restrictive in order to remove false positive candidates. The strictness can be reduced by using the <code class="docutils literal"><span class="pre">--lenient</span></code> option.</li>
<li>Leverage bcolz indexing for the built-in inheritance model tools to dramatically improve speed.</li>
<li>Support for multi-generational pedigrees for the built in inheritance model tools. (thanks to Jessica, Andrew,
and jmcelwee for extensive discussion <a class="reference external" href="https://github.com/arq5x/gemini/issues/388">https://github.com/arq5x/gemini/issues/388</a>)</li>
<li>Leverage genotype likelihoods in tools other than <code class="docutils literal"><span class="pre">mendel_errors</span></code> as a means to filter variants.</li>
<li>Automatically phase genotypes by transmission on the fly for the <cite>comp_hets</cite> tool.</li>
<li>Further performance improvements for bcolz queries</li>
<li>The <code class="docutils literal"><span class="pre">--affected-only</span></code> option has been made the default and it&#8217;s opposing replacement named <code class="docutils literal"><span class="pre">--allow-unaffected</span></code> to revert.</li>
<li>Fixed a reporting error for the inheritance tools (i.e., family_id was mis-specified in output).</li>
<li>Annotate the variants table with impact even if there is not severe impact. Thanks to &#64;mjsduncan for reporting.</li>
<li>Reduce memory requirements when loading. Thanks to &#64;mjsduncan for reporting.</li>
</ol>
</div>
<div class="section" id="id10">
<h2>0.15.1<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Fix regression in grabix. Thanks to Sven-Eric Shelhorn for reporting.</li>
<li>Fix handling of samples with &#8220;-&#8221;. Thanks to Uma Paila for reporting.</li>
</ol>
</div>
<div class="section" id="id11">
<h2>0.15.0<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Use external index to speed genotype queries (this is created by default on load unless &#8211;no-bcolz is specified)</li>
<li>Match on ref and alternate alleles (not just position) when annotating with VCF. Thanks Jeremy Goecks.</li>
<li>Related to matching, we now load extra annotation, e.g. VEP as VCF and require ref and alt matching. Previously was done with bed overlap.</li>
<li>Faster queries due to lazy loading of genotype columns.</li>
<li>Read gt* columns from the database for better backward compatibility.</li>
<li>Code cleanup. Thanks to Christian Brueffer.</li>
</ol>
</div>
<div class="section" id="id12">
<h2>0.14.0<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Standardized the output from the built-in tools into a common, BED+ format. Thanks to feedback from Jessica Chong and Daniel Gaston.</li>
<li>Release of <cite>mendel_errors</cite> tool which also outputs the type of error and the probability (based on PL&#8217;s)</li>
<li>Improvements to the <cite>load</cite> tool when running on large compute clusters using PBS, SGE, SLURM, etc. Also provde a workaround for NFS locking issues. Many thanks to Ben Weisburd in Daniel Macarthur&#8217;s lab.</li>
<li>Improve preprocess script to support varscan, platypus (<a class="reference external" href="https://gist.github.com/brentp/4db670df147cbd5a2b32">https://gist.github.com/brentp/4db670df147cbd5a2b32</a>)</li>
<li>Performance improvements for many of the built-in tools (pre-compile evals)</li>
<li>Bug fix for installation with sudo privileges.</li>
</ol>
</div>
<div class="section" id="apr-09">
<h2>0.13.1 (2015-Apr-09)<a class="headerlink" href="#apr-09" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Major <cite>query</cite> speed improvements. For example, the following query goes from 43 seconds in version 0.12.2 to 11 seconds in 0.13.0. All queries involving <cite>gt_*</cite> fields should be substantially faster.</li>
</ol>
<blockquote>
<div><div class="highlight-python"><div class="highlight"><pre>$ gemini query \
        -q &quot;select chrom, start, (gts).(*) from variants&quot; data/tmaster.db \
        --gt-filter &quot;(gt_depths).(*).(&gt;=20).(all)&quot; &gt; /dev/null
</pre></div>
</div>
</div></blockquote>
<ol class="arabic simple" start="2">
<li>Speed improvements to <cite>load</cite>. The following went from 7 minutes 9 seconds to 6 minutes 21 seconds.</li>
</ol>
<blockquote>
<div><div class="highlight-python"><div class="highlight"><pre>$ gemini load -t VEP -v data/v100K.vcf.gz data/tmaster.db --cores 4
</pre></div>
</div>
</div></blockquote>
<ol class="arabic simple" start="3">
<li>We added the <cite>gt_phred_ll_homref</cite>, <cite>gt_phred_ll_het</cite>, <cite>gt_phred_ll_homalt</cite> columns to database. These are the genotype likelihoods pulled from the GL or PL columns of the VCF if available. They can all be queried and filtered in the same way as existing gt_* columns. In future releases, we are planning tp use genotype likelihood to assign likelihoods to de novo mutations, mendelian violations, and variants meeting other inheritance patterns.</li>
<li>Fixed bugs related to splitting multiple alts (thanks to &#64;jdh237)</li>
<li>We are working to improve development and release testing. This is ongoing, but we now support gemini_install.py &#8211;version unstable so that users can try out the latest changes and help with testing before releases. gemini_update is still limited to master as the most recent version.</li>
<li>Update cyvcf so it doesn&#8217;t error when AD tag is used for non-list data.</li>
<li>Fix regression in cyvcf to handle Flags in info field. (Thanks to Jon for reporting)</li>
<li>Improvements to install related to PYTHONHOME and other env variables(&#64;chapmanb &amp; &#64;bw2)</li>
</ol>
</div>
<div class="section" id="id13">
<h2>0.12.2<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h2>
<p>Corrected a stale .c file in the cyvcf library. This is effectively a replacement for the 0.12.1 release.</p>
</div>
<div class="section" id="id14">
<h2>0.12.1<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Support for input VCF files containing variants with multiple alternate alleles. Thanks to Brent Pedersen.</li>
<li>Updated, decomposed, and normalized the ExAC, Clinvar, Cosmic, dbSNP, and ESP annotation files to properly support variants with multiple alternate alleles.</li>
<li>Updated the logic for the clinvar significance column to retain all documented singificances.</li>
<li>Support for VCF annotation files in the <cite>annotate</cite> tool.</li>
<li>Improved the speed of loading by 10-15%. Thanks to Brent Pedersen.</li>
<li>Added <cite>&#8211;only-affected</cite> and <cite>&#8211;min-kindreds</cite> options to the compound heterozygotes tool.</li>
<li>Added a <cite>&#8211;format vcf</cite> option to the <cite>query</cite> tool to output query results in VCF format.</li>
<li>Added the <cite>&#8211;families</cite> option to the <cite>auto_*</cite>, <cite>de_novo</cite>, and <cite>comp_hets</cite> tools. Thanks to Mark Cowley and Tony Roscioli.</li>
<li>Added the <cite>&#8211;only-affected</cite> option to the <cite>de_novo</cite> tool.</li>
<li>Allow the <cite>&#8211;sample-filter</cite> to work with <cite>&#8211;format TPED</cite>. Thanks to Rory Kirchner.</li>
<li>Add <cite>&#8211;format sampledetail</cite> option that provides a melted/tidy/flattened version of samples along with <cite>&#8211;showsample</cite> and includes information from samples table. Thanks to Brad Chapman.</li>
<li>Add &#8216;not&#8217; option to &#8211;in filtering. Thanks to Rory Kirchner.</li>
<li>Fixed a bug in the <cite>de_novo</cite> tool that prevented proper function when families have affected and unaffected children. Thanks to Andrew Oler.</li>
<li>Fixed a bug in cyvcf that falsely treated &#8216;.|.&#8217; genotypes as homozygous alternate.  Thanks to Xiao Xu.</li>
<li>GEMINI now checks for and warns of old grabix index files. Thanks to Andrew Oler and Brent Pedersen.</li>
<li>Fixed a bug that added newlines at the end of tab delimited PED files.  Thanks to Brad Chapman.</li>
</ol>
</div>
<div class="section" id="id15">
<h2>0.11.0<a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Integration of ExAC annotations (v0.2): <a class="reference external" href="http://exac.broadinstitute.org/">http://exac.broadinstitute.org/</a></li>
<li>New tools for cancer genome analysis. Many thanks to fantastic work from Colby Chiang.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><cite>gemini set_somatic</cite></li>
<li><cite>gemini actionable_mutations</cite></li>
<li><cite>gemini fusions</cite></li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="3">
<li>Improved support for structural variants. New columns include:</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><cite>sv_cipos_start_left</cite></li>
<li><cite>sv_cipos_end_left</cite></li>
<li><cite>sv_cipos_start_right</cite></li>
<li><cite>sv_cipos_end_right</cite></li>
<li><cite>sv_length</cite></li>
<li><cite>sv_is_precise</cite></li>
<li><cite>sv_tool</cite></li>
<li><cite>sv_evidence_type</cite></li>
<li><cite>sv_event_id</cite></li>
<li><cite>sv_mate_id</cite></li>
<li><cite>sv_strand</cite></li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="4">
<li>Updated the 1000 Genomes annotations to the Phase variant set.</li>
<li>Added <cite>clinvar_causal_allele</cite> column.</li>
<li>Fixed a bug in grabix that caused occasional duplicate and missed variants.</li>
</ol>
</div>
<div class="section" id="id16">
<h2>0.10.1<a class="headerlink" href="#id16" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Add <cite>fitCons &lt;http://biorxiv.org/content/early/2014/09/11/006825&gt;</cite> scores as
an additional measure of potential function in variants of interest, supplementing
existing CADD and dbNSFP approaches.</li>
<li>Updated Clinvar, COSMIC, and dbSNP to their latest versions.</li>
</ol>
</div>
<div class="section" id="id17">
<h2>0.10.0<a class="headerlink" href="#id17" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Provide an <code class="docutils literal"><span class="pre">--annotation-dir</span></code> argument that specifies the path the
annotation databases, to overwrite configured data inputs. Thanks to Björn Grüning,</li>
<li>Support reproducible versioned installs of GEMINI with Python
dependencies. Enables Galaxy integration. Thanks to Björn Grüning,</li>
</ol>
</div>
<div class="section" id="id18">
<h2>0.8.0<a class="headerlink" href="#id18" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Support arbitrary annotation supplied to VEP, which translate into queryable
columns in the main variant table.</li>
<li>Improve the power of the genotype filter wildcard functionality.</li>
</ol>
</div>
<div class="section" id="id19">
<h2>0.7.1<a class="headerlink" href="#id19" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Suppress openpyxl/pandas warnings (thanks to &#64;chapmanb)</li>
<li>Fix unit tests to account for cases where a user has not downloaded the CADD or GERP annotation files.
Thanks to Xialoin Zhu and Daniel Swensson for reporting this and to Uma Paila for correcting it.</li>
</ol>
</div>
<div class="section" id="id20">
<h2>0.7.0<a class="headerlink" href="#id20" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Added support for CADD scores via new <code class="docutils literal"><span class="pre">cadd_raw</span></code> and <code class="docutils literal"><span class="pre">cadd_scaled</span></code> columns.</li>
<li>Added support for genotype wildcards in query select statements. E.g., <code class="docutils literal"><span class="pre">SELECT</span> <span class="pre">chrom,</span> <span class="pre">start,</span> <span class="pre">end</span> <span class="pre">(gts).(phenotype==2)</span> <span class="pre">FROM</span> <span class="pre">variants</span></code>. See <a class="reference external" href="http://gemini.readthedocs.org/en/latest/content/querying.html#selecting-sample-genotypes-based-on-wildcards">http://gemini.readthedocs.org/en/latest/content/querying.html#selecting-sample-genotypes-based-on-wildcards</a>.</li>
<li>Added support for genotype wildcards in the &#8211;gt-filter. E.g., <code class="docutils literal"><span class="pre">--gt-filter</span> <span class="pre">&quot;(gt_types).(phenotype==2).(==HET)</span></code>. See <a class="reference external" href="http://gemini.readthedocs.org/en/latest/content/querying.html#gt-filter-wildcard-filtering-on-genotype-columns">http://gemini.readthedocs.org/en/latest/content/querying.html#gt-filter-wildcard-filtering-on-genotype-columns</a>.</li>
<li>Added support for the VCF INFO field both in the API and as a column that can be SELECT&#8217;ed.</li>
<li>Upgraded to the latest version of ClinVar.</li>
<li>Standardized impacts to use Sequence Ontology (SO) terms.</li>
<li>Automatically add indexes to custom, user-supplied annotation columns.</li>
<li>Improvements to the installation script.</li>
<li>Fixed bugs in the handling of ClinVar UTF8 encoded strings.</li>
<li>Upgraded the <code class="docutils literal"><span class="pre">gene_summary</span></code> and <code class="docutils literal"><span class="pre">gene_detailed</span></code> tables to version 75 of Ensembl.</li>
<li>Added support for the MPI Mouse Phenotype database via the <code class="docutils literal"><span class="pre">mam_phenotype_id</span></code> column in the <code class="docutils literal"><span class="pre">gene_summary</span></code> table.</li>
<li>Enhanced security.</li>
<li>Corrected the ESP allele frequencies to be based report _alternate_ allele frequency instead of _minor_ allele frequency.</li>
<li>VEP version support updated (73-75) Support for aa length and bio type in VEP.</li>
<li>The <cite>lof_sieve</cite> tool support has been extended to VEP annotations.</li>
<li>Added the <code class="docutils literal"><span class="pre">ccds_id</span></code> and <code class="docutils literal"><span class="pre">entrez_id</span></code> columns to the <code class="docutils literal"><span class="pre">gene_detailed</span></code> table.</li>
</ol>
</div>
<div class="section" id="id21">
<h2>0.6.6<a class="headerlink" href="#id21" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Added COSMIC mutation information via new cosmic_ids column.</li>
</ol>
</div>
<div class="section" id="jan-03">
<h2>0.6.4 (2014-Jan-03)<a class="headerlink" href="#jan-03" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>New annotation: experimentally validated human enhancers from VISTA.</li>
<li>Installation improvements to enable isolated installations inside of virtual
machines and containers without data. Allow data-only upgrades as part of
<code class="docutils literal"><span class="pre">update</span></code> process.</li>
<li>Fix for gemini query error when <code class="docutils literal"><span class="pre">--header</span></code> specified (#241).</li>
</ol>
</div>
<div class="section" id="dec-10">
<h2>0.6.3.2 (2013-Dec-10)<a class="headerlink" href="#dec-10" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Fixed a bug that caused <code class="docutils literal"><span class="pre">--gt-filter</span></code> to no be enforced from <code class="docutils literal"><span class="pre">query</span></code> tool unless a GT* column was selected.</li>
<li>Support for ref and alt allele depths provided by FreeBayes.</li>
</ol>
</div>
<div class="section" id="nov-19">
<h2>0.6.3.1 (2013-Nov-19)<a class="headerlink" href="#nov-19" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Fixed undetected bug preventing the <code class="docutils literal"><span class="pre">comp_hets</span></code> tool from functioning.</li>
<li>Added unit tests for the <code class="docutils literal"><span class="pre">comp_hets</span></code> tool.</li>
</ol>
</div>
<div class="section" id="nov-7">
<h2>0.6.3 (2013-Nov-7)<a class="headerlink" href="#nov-7" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Addition permutation testing to the c-alpha test via the <code class="docutils literal"><span class="pre">--permutations</span></code>
option.</li>
<li>Addition of the <code class="docutils literal"><span class="pre">--passonly</span></code> option during loading to filter out all
variants with a filter flag set.</li>
<li>Fixed bug with parallel loading using the extended sample table format.</li>
<li>SLURM support added.</li>
<li>Refactor of loading options to remove explosion of xxx-queue options. Now
load using <code class="docutils literal"><span class="pre">--scheduler</span></code> on <code class="docutils literal"><span class="pre">--queue</span></code>.</li>
<li>Refactor of Sample class to handle the expanded samples table.</li>
<li>Addition of <code class="docutils literal"><span class="pre">--carrier-summary-by-phenotype</span></code> for summarizing the counts of
carriers and non-carriers stratified by the given sample phenotype column.</li>
<li>Added a <code class="docutils literal"><span class="pre">--nonsynonymous</span></code> option to the C-alpha test.</li>
<li>Added <code class="docutils literal"><span class="pre">gemini</span> <span class="pre">amend</span></code> to edit an existing database. For now only handles updating
the samples table.</li>
<li>Fixed a bug that prevented variants that overlapped with multiple 1000G variants
from having AAF info extracted from 1000G annotations.  This is now corrected such
that multiple overlaps with 1000G variants are tolerated, yet the logic ensures
that the AAF info is extracted for the correct variant.</li>
<li>Fixed installation issues for the GEMINI browser.</li>
<li>Added <code class="docutils literal"><span class="pre">--show-families</span></code> option to gemini query.</li>
</ol>
</div>
<div class="section" id="oct-7">
<h2>0.6.2 (2013-Oct-7)<a class="headerlink" href="#oct-7" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Moved <cite>&#8211;tped</cite> and <cite>&#8211;json</cite> options into the more generic <cite>&#8211;format</cite> option.</li>
<li>Fixed bug in handling missing phenotypes in the sample table.</li>
<li>Fixed <cite>&#8211;tped</cite> output formatting error.</li>
<li>API change: GeminiQuery.run takes an optional list of predicates that a row
must pass to be returned.</li>
<li><cite>&#8211;sample-filter</cite> option added to allow for restricting variants to samples
that pass the given sample query.</li>
<li>ethnicity removed as a default PED field.</li>
<li>PED file format extended to allow for extra columns to be added to the samples table under the column named in the header.</li>
<li>The autosomal_recessive and autosomal_dominant tools now warn, but allow for variants to be detected in the absence of known parent/child relationships.</li>
</ol>
</div>
<div class="section" id="sep-09">
<h2>0.6.1 (2013-Sep-09)<a class="headerlink" href="#sep-09" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Corrected bug in de_novo tool that was undetected in 0.6.0.  Unit tests have been added to head this off in the future. Thanks to <strong>Jessica Chong</strong></li>
<li>Added the <cite>-d</cite> option (minimum sequence depth allowed for a genotype) to the <cite>autosmal_recessive</cite> and <cite>autosmal_dominant</cite> tools.</li>
<li>New <cite>&#8211;tped</cite> option in the <cite>query</cite> tool for reporting variants in TPED format. Thanks to <strong>Rory Kirchner</strong>.</li>
<li>New <cite>&#8211;tfam</cite> option in the <cite>dump</cite> tool for reporting sample infor in TFAM format. Thanks to <strong>Rory Kirchner</strong>.</li>
</ol>
</div>
<div class="section" id="sep-02">
<h2>0.6.0 (2013-Sep-02)<a class="headerlink" href="#sep-02" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Add the <code class="docutils literal"><span class="pre">--min-kindreds</span></code> option to the <code class="docutils literal"><span class="pre">autosomal_recessive</span></code> and <code class="docutils literal"><span class="pre">autosomal_dominant</span></code> tools to restrict candidate variants/genes to those affecting at least <code class="docutils literal"><span class="pre">--min-kindreds</span></code>. Thanks to <strong>Jessica Chong</strong></li>
<li>Addition of a new <code class="docutils literal"><span class="pre">burden</span></code> tool for gene or region based burden tests.  First release supports the C-alpha test.  Thanks to <strong>Rory Kirchner</strong>.</li>
<li>Use of Continuum Analytics Anaconda python package for the automated installer. Thanks to <strong>Brad Chapman</strong>.</li>
<li>Enhancements to the <code class="docutils literal"><span class="pre">annotate</span></code> tool allowing one to create new database columns from values in custom BED+ annotation files.  Thanks to <strong>Jessica Chong</strong> and <strong>Graham Ritchie</strong>.</li>
<li>Addition of the <code class="docutils literal"><span class="pre">--column</span></code>, <code class="docutils literal"><span class="pre">--filter</span></code>, and <code class="docutils literal"><span class="pre">--json</span></code> options to the <code class="docutils literal"><span class="pre">region</span></code> tool.</li>
<li>Improvements to unit tests.</li>
<li>Allow alternate sample delimiters in the <code class="docutils literal"><span class="pre">query</span></code> tool via the <code class="docutils literal"><span class="pre">--sample-delim</span></code> option.  Thanks to <strong>Jessica Chong</strong>.</li>
<li>Provide a REST-like interface to the gemini browser.  In support of future visualization tools.</li>
<li>Allow the <code class="docutils literal"><span class="pre">query</span></code> tool to report results in JSON format via the <code class="docutils literal"><span class="pre">--json</span></code> option.</li>
<li>Various minor improvements and bug fixes.</li>
</ol>
</div>
<div class="section" id="b-2013-jul-23">
<h2>0.5.0b (2013-Jul-23)<a class="headerlink" href="#b-2013-jul-23" title="Permalink to this headline">¶</a></h2>
<ol class="arabic">
<li><p class="first">Tolerate either -9 or 0 for unknown parent or affected status in PED files.</p>
</li>
<li><p class="first">Refine the rules for inheritance and parental affected status for autosomal dominant inheritance models.</p>
</li>
<li><p class="first">The <code class="docutils literal"><span class="pre">autosomal_dominant</span></code>, <code class="docutils literal"><span class="pre">autosomal_recessive</span></code>, and <code class="docutils literal"><span class="pre">de_novo</span></code> mutation tools have received the following improvements.</p>
<blockquote>
<div><ul class="simple">
<li>improved speed (especially when there are multiple families)</li>
<li>by default, all columns in the variant table are reported and no conditions are placed on the returned variants.  That is, as long as the variant meets
the inheritance model, it will be reported.</li>
<li>the addition of a <code class="docutils literal"><span class="pre">--columns</span></code> option allowing one to override the above default behavior and report a subset of columns.</li>
<li>the addition of a <code class="docutils literal"><span class="pre">--filter</span></code> option allowing one to override the above default behavior and filter reported variants based on specific criteria.</li>
</ul>
</div></blockquote>
</li>
</ol>
<p>4. The default minimum aligned sequencing depth for each variant reported by
the <code class="docutils literal"><span class="pre">de_novo</span></code> tool is 0.  Greater stringency can be applied with the <code class="docutils literal"><span class="pre">-d</span></code>
option.</p>
</div>
<div class="section" id="b-2013-jun-12">
<h2>0.4.0b (2013-Jun-12)<a class="headerlink" href="#b-2013-jun-12" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Added new <code class="docutils literal"><span class="pre">gt_ref_depths</span></code>, <code class="docutils literal"><span class="pre">gt_alt_depths</span></code>, and <code class="docutils literal"><span class="pre">gt_quals</span></code> columns.</li>
<li>Added a new <code class="docutils literal"><span class="pre">--show-samples</span></code> option to the <code class="docutils literal"><span class="pre">query</span></code> module to display samples with alternate allele genotypes.</li>
<li>Improvements and bug fixes for installation.</li>
</ol>
</div>
<div class="section" id="b">
<h2>0.3.0b<a class="headerlink" href="#b" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Improved speed for adding custom annotations.</li>
<li>Added GERP conserved elements.</li>
<li>Optionally addition of GERP conservation scores at base pair resolution.</li>
<li>Move annotation files to Amazon S3.</li>
</ol>
</div>
</div>




  

<div id="disqus_thread"></div>
 <script type="text/javascript">
     /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
     var disqus_shortname = 'geminidocs'; // required: replace example with your forum shortname

     /* * * DON'T EDIT BELOW THIS LINE * * */
     (function() {
         var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
         dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
         (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
     })();
 </script>
 <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
 <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>






          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/gemini.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Release History</a><ul>
<li><a class="reference internal" href="#future">0.19.0 (future)</a></li>
<li><a class="reference internal" href="#id1">0.18.1</a></li>
<li><a class="reference internal" href="#id2">0.18.0</a></li>
<li><a class="reference internal" href="#id3">0.17.2</a></li>
<li><a class="reference internal" href="#id4">0.17.1</a></li>
<li><a class="reference internal" href="#id5">0.17.0</a></li>
<li><a class="reference internal" href="#id6">0.16.3</a></li>
<li><a class="reference internal" href="#id7">0.16.2</a></li>
<li><a class="reference internal" href="#id8">0.16.1</a></li>
<li><a class="reference internal" href="#id9">0.16.0</a></li>
<li><a class="reference internal" href="#id10">0.15.1</a></li>
<li><a class="reference internal" href="#id11">0.15.0</a></li>
<li><a class="reference internal" href="#id12">0.14.0</a></li>
<li><a class="reference internal" href="#apr-09">0.13.1 (2015-Apr-09)</a></li>
<li><a class="reference internal" href="#id13">0.12.2</a></li>
<li><a class="reference internal" href="#id14">0.12.1</a></li>
<li><a class="reference internal" href="#id15">0.11.0</a></li>
<li><a class="reference internal" href="#id16">0.10.1</a></li>
<li><a class="reference internal" href="#id17">0.10.0</a></li>
<li><a class="reference internal" href="#id18">0.8.0</a></li>
<li><a class="reference internal" href="#id19">0.7.1</a></li>
<li><a class="reference internal" href="#id20">0.7.0</a></li>
<li><a class="reference internal" href="#id21">0.6.6</a></li>
<li><a class="reference internal" href="#jan-03">0.6.4 (2014-Jan-03)</a></li>
<li><a class="reference internal" href="#dec-10">0.6.3.2 (2013-Dec-10)</a></li>
<li><a class="reference internal" href="#nov-19">0.6.3.1 (2013-Nov-19)</a></li>
<li><a class="reference internal" href="#nov-7">0.6.3 (2013-Nov-7)</a></li>
<li><a class="reference internal" href="#oct-7">0.6.2 (2013-Oct-7)</a></li>
<li><a class="reference internal" href="#sep-09">0.6.1 (2013-Sep-09)</a></li>
<li><a class="reference internal" href="#sep-02">0.6.0 (2013-Sep-02)</a></li>
<li><a class="reference internal" href="#b-2013-jul-23">0.5.0b (2013-Jul-23)</a></li>
<li><a class="reference internal" href="#b-2013-jun-12">0.4.0b (2013-Jun-12)</a></li>
<li><a class="reference internal" href="#b">0.3.0b</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="caveats.html"
                        title="previous chapter">&lt;no title&gt;</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="faq.html"
                        title="next chapter">F.A.Q.</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/content/history.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="faq.html" title="F.A.Q."
             >next</a> |</li>
        <li class="right" >
          <a href="caveats.html" title="&lt;no title&gt;"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">gemini v0.18.1</a> &raquo;</li> 
      </ul>
    </div>

    <div class="footer">
        &copy; Copyright 2012,2013,2014,2015.
      Last updated on Feb 05, 2016.
    <a href="http://www.gnu.org/licenses/gpl-3.0.txt"><p>GPL3 licensed</p></a>
    </div>

<!-- 
<div id="editor-trap">
      <h3>Edit and improve this document!</h3>

      <p>
        This file can be edited directly through the Web. Anyone can
        update and fix errors in this document with few clicks --
        no downloads needed.
      <p>

      <ol>

        <li>
          Go to
        <a href="https://github.com/arq5x/gemini/blob/master/docs/content/history.rst">
          Release History
        </a> on GitHub.
      </li>

        <li>
        <b>Edit</b> files using GitHub's text editor in your web browser (see the 'Edit' tab on the top right of the file)
      </li>

      <li>
        Fill in the <b>Commit message</b> text box at the bottom of the page describing <i>why</i>
        you made the changes. Press the <b>Propose file change</b> button next to it when done.
      </li>

      <li>
        Then click <b>Send a pull request</b>.
      </li>

      <li>
        Your changes are now queued for review under the project's <a href="https://github.com/arq5x/gemini/pulls">Pull requests</a> tab on GitHub!
      </li>
      </ol>

      <p>
        For an introduction to the documentation format please see <a href="http://docutils.sourceforge.net/docs/user/rst/quickstart.html">the reST primer</a>.
      </p>

  </div>
 -->
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', UA-39150563-1]);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>


  </body>
</html>